submissions/hornwitser_dissector/wireshark-dissector.md

```diff
@@ -0,0 +1,49 @@
+And so it became that this project was put on shelf not to be seen again.
+
+In the past _Twinsen_ has posted about making a dissector for Wireshark in [FFF #196](https://factorio.com/blog/post/fff-196).
+The post contains quite a lot of internal information about the protocol which I could have used when I made my dissector, but I chose not to as I wanted to make mine without relying on internal information from the game source code.
+An obvious question right now though is why not use _Twinsen_'s dissector instead of writing my own?
+Well for a start it's not publicly available.
+But more importantly _it can't be made publicly available_.
+The way _Twinsen's_ dissector worked was to build Factorio as a shared library, and then build a custom version of Wireshark that linked to that shared library.
+But Wireshark is licensed under the GPL and you are not allowed to distribute GPL code that links with proprietary code, so to distribute the dissector Wube would have to release Factorio under the GPL which is obviously not going to happen.
+
+But a few weeks ago, I realized something.
+The Windows version of Factorio ships with a .pdb file.
+It's stuffed with debug symbols, the kind of data you'd need to attach a debugger to Factorio and inspect the values of variables used in the code.
+And that means it contains type information.
+Type information means the structure of the data as defined in the code along with the names of fields and enums.
+The name of fields are very useful as it's the label developers put on a piece of data in order to reference it in the code.
+And likewise the enums contain mappings of logical names to numeric values.
+
+I was able to dump this type information from the .pdb file into something readable using the cvdump.exe tool.
+And once I had that it was pretty easy to find the names of the structures from the .pdb file rename all of the fields in my dissector to use them.
+I could also get the enum definitions to translate numeric values to logical names for the fields that use them.
+And after some tidying up of the code and how the data was presented I was able to make the dissector output look like this:
+
+![](media/new-dissector.png)
+
+No longer does one have to guess what an "Odd Blue" value is, it's clearly it's the id of an InputActionSegment.
+And you can alse see the type is translated from the machine code looking 0xa2 into the much more friendly name "ServerCommand".
+But by doing this my dissector is no longer a cleanly separate thing from Factorio and how that affects the legallites around distributing this dissector I really have no clue.
+I asked the Factorio developers and they said they were fine with me releasing this to the public.
+So if you want to try it out yourself or add it to your toolbox you can find it [in my repository for it on GitHub](https://github.com/Hornwitser/factorio_dissector)
```

Comment:
In closing you could mention what you achieved with your tool. Did you find the issue why players were dropping, or did upi only figure it out too long afterwards? Or was it just a challenge in the end? And what can players achieve or learn by using this dissector?

Reply:
I tried to write something in the end for this. I'm not feeling like it fit well though.

Reply:
I think it came out good enough. If you're not happy with it you can always remove it again. Therenas will likely tweak it a little anyway.